src/cmd-sign

```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/python3
+
+import argparse
+import gi
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+import tarfile
+import time
+
+import botocore
+import boto3
+
+from fedora_messaging import message
+from fedora_messaging.api import publish
+from fedora_messaging.config import conf
+
+sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
+from cmdlib import load_json, get_basearch, Builds, import_ostree_commit  # noqa: E402
+
+gi.require_version('OSTree', '1.0')
+from gi.repository import Gio, OSTree
+
+FEDORA_MESSAGING_TOPIC_PREFIX = "org.fedoraproject.prod.coreos.build.request"
+
+
+def main():
+    args = parse_args()
+    args.func(args)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--build", help="Build ID", default='latest')
+    subparsers = parser.add_subparsers(dest='cmd', title='subcommands')
+    subparsers.required = True
+
+    robosig = subparsers.add_parser('robosignatory', help='sign with '
+                                    'RoboSignatory via fedora-messaging')
+    robosig.add_argument("--s3", metavar='<BUCKET>[/PREFIX]', required=True,
+                         help="Bucket and prefix to S3 builds/ dir")
+    group = robosig.add_mutually_exclusive_group(required=True)
+    group.add_argument("--ostree", help="Sign commit", action='store_true')
+    group.add_argument("--images", help="Sign images", action='store_true')
+    robosig.add_argument("--extra-fedmsg-keys", action='append',
+                         metavar='KEY=VAL', default=[])
+    robosig.add_argument("--fedmsg-conf", metavar='CONFIG.TOML')
+    robosig.add_argument("--gpgkeypath", help="Path to directory containing "
+                         "public keys to use for signature verification",
+                         default="/etc/pki/rpm-gpg")
+    robosig.set_defaults(func=cmd_robosignatory)
+
+    return parser.parse_args()
+
+
+def cmd_robosignatory(args):
+    builds = Builds()
+    if args.build == 'latest':
+        args.build = builds.get_latest()
+
+    s3 = boto3.client('s3')
+    args.bucket, args.prefix = get_bucket_and_prefix(args.s3)
+
+    args.extra_keys = {}
+    for keyval in args.extra_fedmsg_keys:
+        key, val = keyval.split('=', 1)  # will throw if there's no =
+        args.extra_keys[key] = val
+
+    if args.fedmsg_conf:
+        conf.load_config(args.fedmsg_conf)
+
+    # these two are different enough that they deserve separate handlers
+    if args.ostree:
+        robosign_ostree(args, s3)
+    else:
+        assert args.images
+        robosign_images(args, s3)
+
+
+def robosign_ostree(args, s3):
+    builds = Builds()
+    builddir = builds.get_build_dir(args.build)
+    build = load_json(f'{builddir}/meta.json')
+    checksum = build['ostree-commit']
+
+    # Copy commit object to a temporary location. Another approach here is to
```

Comment:
hmm. seems like this is a bit messy. I guess maybe we can change it to say "A preferred approach". It would be nice if we go back later and look at the messages we are able to use the same bucket/key to download the commit object (i.e. if it were the final location in the bucket).